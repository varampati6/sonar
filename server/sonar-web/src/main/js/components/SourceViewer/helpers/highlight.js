/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import escapeHtml from 'escape-html';
import { uniq } from 'lodash';

export type Token = { className: string, markers: Array<number>, text: string };
export type Tokens = Array<Token>;

const ISSUE_LOCATION_CLASS = 'source-line-code-issue';

export const splitByTokens = (code: string, rootClassName: string = ''): Tokens => {
  const container = document.createElement('div');
  let tokens = [];
  container.innerHTML = code;
  [].forEach.call(container.childNodes, node => {
    if (node.nodeType === 1) {
      // ELEMENT NODE
      const fullClassName = rootClassName ? rootClassName + ' ' + node.className : node.className;
      const innerTokens = splitByTokens(node.innerHTML, fullClassName);
      tokens = tokens.concat(innerTokens);
    }
    if (node.nodeType === 3) {
      // TEXT NODE
      tokens.push({ className: rootClassName, markers: [], text: node.nodeValue });
    }
  });
  return tokens;
};

export const highlightSymbol = (tokens: Tokens, symbol: string): Tokens => {
  const symbolRegExp = new RegExp(`\\b${symbol}\\b`);
  return tokens.map(
    token =>
      (symbolRegExp.test(token.className)
        ? { ...token, className: `${token.className} highlighted` }
        : token)
  );
};

/**
 * Intersect two ranges
 * @param s1 Start position of the first range
 * @param e1 End position of the first range
 * @param s2 Start position of the second range
 * @param e2 End position of the second range
 */
const intersect = (
  s1: number,
  e1: number,
  s2: number,
  e2: number
): { from: number, to: number } => {
  return { from: Math.max(s1, s2), to: Math.min(e1, e2) };
};

/**
 * Get the substring of a string
 * @param str A string
 * @param from "From" offset
 * @param to "To" offset
 * @param acc Global offset to eliminate
 */
const part = (str: string, from: number, to: number, acc: number): string => {
  // we do not want negative number as the first argument of `substr`
  return from >= acc ? str.substr(from - acc, to - from) : str.substr(0, to - from);
};

/**
 * Highlight issue locations in the list of tokens
 */
export const highlightIssueLocations = (
  tokens: Tokens,
  issueLocations: Array<*>,
  rootClassName: string = ISSUE_LOCATION_CLASS
): Tokens => {
  issueLocations.forEach(location => {
    const nextTokens = [];
    let acc = 0;
    let markerAdded = location.line !== location.startLine;
    tokens.forEach(token => {
      const x = intersect(acc, acc + token.text.length, location.from, location.to);
      const p1 = part(token.text, acc, x.from, acc);
      const p2 = part(token.text, x.from, x.to, acc);
      const p3 = part(token.text, x.to, acc + token.text.length, acc);
      if (p1.length) {
        nextTokens.push({ ...token, text: p1 });
      }
      if (p2.length) {
        const newClassName = token.className.indexOf(rootClassName) === -1
          ? `${token.className} ${rootClassName}`
          : token.className;
        nextTokens.push({
          className: newClassName,
          markers: !markerAdded && location.index != null
            ? uniq([...token.markers, location.index])
            : token.markers,
          text: p2
        });
        markerAdded = true;
      }
      if (p3.length) {
        nextTokens.push({ ...token, text: p3 });
      }
      acc += token.text.length;
    });
    tokens = nextTokens.slice();
  });
  return tokens;
};

export const generateHTML = (tokens: Tokens): string => {
  return tokens
    .map(token => `<span class="${token.className}">${escapeHtml(token.text)}</span>`)
    .join('');
};
