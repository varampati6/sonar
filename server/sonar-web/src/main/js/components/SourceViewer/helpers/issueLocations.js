/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import type { TextRange, Issue } from '../../issue/types';

export const getLinearLocations = (
  textRange?: TextRange
): Array<{ line: number, from: number, to: number }> => {
  if (!textRange) {
    return [];
  }
  const locations = [];

  // go through all lines of the `textRange`
  for (let line = textRange.startLine; line <= textRange.endLine; line++) {
    // TODO fix 999999
    const from = line === textRange.startLine ? textRange.startOffset : 0;
    const to = line === textRange.endLine ? textRange.endOffset : 999999;
    locations.push({ line, from, to });
  }
  return locations;
};

export const getIssueLocations = (
  issue: Issue
): Array<{
  msg: string,
  flowIndex: number,
  locationIndex: number,
  textRange?: TextRange,
  index?: number
}> => {
  const allLocations = [];
  issue.flows.forEach((locations, flowIndex) => {
    if (locations) {
      const locationsCount = locations.length;
      locations.forEach((location, index) => {
        const flowLocation = {
          ...location,
          flowIndex,
          locationIndex: index,
          // set index only for real flows, do not set for just secondary locations
          index: locationsCount > 1 ? locationsCount - index : undefined
        };
        allLocations.push(flowLocation);
      });
    }
  });
  return allLocations;
};
