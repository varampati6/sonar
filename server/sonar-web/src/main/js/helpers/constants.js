/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
export const SEVERITIES = ['BLOCKER', 'CRITICAL', 'MAJOR', 'MINOR', 'INFO'];
export const STATUSES = ['OPEN', 'REOPENED', 'CONFIRMED', 'RESOLVED', 'CLOSED'];

export const CHART_COLORS_RANGE_PERCENT = ['#00aa00', '#b0d513', '#eabe06', '#ed7d20', '#d4333f'];
export const CHART_REVERSED_COLORS_RANGE_PERCENT = [
  '#d4333f',
  '#ed7d20',
  '#eabe06',
  '#b0d513',
  '#00aa00'
];

export const RATING_COLORS = ['#00aa00', '#b0d513', '#eabe06', '#ed7d20', '#e00'];
