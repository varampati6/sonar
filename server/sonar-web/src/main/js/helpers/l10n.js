/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/* @flow */
import moment from 'moment';
import { request } from './request';

let messages = {};

export function translate(...keys: string[]) {
  const messageKey = keys.join('.');
  return messages[messageKey] || messageKey;
}

export function translateWithParameters(messageKey: string, ...parameters: Array<string | number>) {
  const message = messages[messageKey];
  if (message) {
    return parameters
      .map(parameter => String(parameter))
      .reduce((acc, parameter, index) => acc.replace(`{${index}}`, parameter), message);
  } else {
    return `${messageKey}.${parameters.join('.')}`;
  }
}

export function hasMessage(...keys: string[]) {
  const messageKey = keys.join('.');
  return messages[messageKey] != null;
}

export function configureMoment(language?: string) {
  moment.locale(language || getPreferredLanguage());
}

function getPreferredLanguage() {
  return window.navigator.languages ? window.navigator.languages[0] : window.navigator.language;
}

function makeRequest(params) {
  const url = '/api/l10n/index';

  return request(url).setData(params).submit().then(response => {
    switch (response.status) {
      case 200:
        return response.json();
      case 304:
        return JSON.parse(localStorage.getItem('l10n.bundle') || '{}');
      case 401:
        window.location =
          window.baseUrl +
          '/sessions/new?return_to=' +
          encodeURIComponent(
            window.location.pathname + window.location.search + window.location.hash
          );
        // return unresolved promise to stop the promise chain
        // anyway the page will be reloaded
        return new Promise(() => {});
      default:
        throw new Error('Unexpected status code: ' + response.status);
    }
  });
}

function checkCachedBundle() {
  const cached = localStorage.getItem('l10n.bundle');

  if (!cached) {
    return false;
  }

  try {
    const parsed = JSON.parse(cached);
    return parsed != null && typeof parsed === 'object';
  } catch (e) {
    return false;
  }
}

export function requestMessages() {
  const currentLocale = getPreferredLanguage();
  const cachedLocale = localStorage.getItem('l10n.locale');
  const params = {};

  if (currentLocale) {
    params.locale = currentLocale;
  }

  if (cachedLocale === currentLocale) {
    const bundleTimestamp = localStorage.getItem('l10n.timestamp');
    if (bundleTimestamp !== null && checkCachedBundle()) {
      params.ts = bundleTimestamp;
    }
  }

  return makeRequest(params).then(response => {
    try {
      const currentTimestamp = moment().format('YYYY-MM-DDTHH:mm:ssZZ');
      localStorage.setItem('l10n.timestamp', currentTimestamp);
      localStorage.setItem('l10n.locale', response.effectiveLocale);
      localStorage.setItem('l10n.bundle', JSON.stringify(response.messages));
    } catch (e) {
      // do nothing
    }
    configureMoment(response.effectiveLocale);
    resetBundle(response.messages);
  });
}

export function resetBundle(bundle: Object) {
  messages = bundle;
}

export function installGlobal() {
  window.t = translate;
  window.tp = translateWithParameters;
  window.requestMessages = requestMessages;
}

export function getLocalizedDashboardName(baseName: string) {
  const l10nKey = `dashboard.${baseName}.name`;
  const l10nLabel = translate(l10nKey);
  return l10nLabel !== l10nKey ? l10nLabel : baseName;
}

export function getLocalizedMetricName(metric: { key: string, name: string }) {
  const bundleKey = `metric.${metric.key}.name`;
  const fromBundle = translate(bundleKey);
  return fromBundle !== bundleKey ? fromBundle : metric.name;
}

export function getLocalizedMetricDomain(domainName: string) {
  const bundleKey = `metric_domain.${domainName}`;
  const fromBundle = translate(bundleKey);
  return fromBundle !== bundleKey ? fromBundle : domainName;
}
