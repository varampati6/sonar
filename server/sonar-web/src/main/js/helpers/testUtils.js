/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
export const mockEvent = {
  target: { blur() {} },
  currentTarget: { blur() {} },
  preventDefault() {},
  stopPropagation() {}
};

export const click = (element, event = {}) => element.simulate('click', { ...mockEvent, ...event });

export const clickOutside = (event = {}) => {
  const dispatchedEvent = new MouseEvent('click', event);
  window.dispatchEvent(dispatchedEvent);
};

export const submit = element =>
  element.simulate('submit', {
    preventDefault() {}
  });

export const change = (element, value) =>
  element.simulate('change', {
    target: { value },
    currentTarget: { value }
  });

export const keydown = keyCode => {
  const event = new KeyboardEvent('keydown', { keyCode });
  document.dispatchEvent(event);
};

export const elementKeydown = (element, keyCode) => {
  element.simulate('keydown', {
    currentTarget: { element },
    keyCode,
    preventDefault() {}
  });
};
