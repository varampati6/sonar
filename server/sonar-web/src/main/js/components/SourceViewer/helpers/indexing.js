/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import { flatten } from 'lodash';
import { splitByTokens } from './highlight';
import { getLinearLocations } from './issueLocations';
import type { Issue } from '../../issue/types';
import type { SourceLine } from '../types';

export type LinearIssueLocation = {
  from: number,
  line: number,
  to: number,
  index?: number
};

export type IndexedIssueLocation = {
  from: number,
  line: number,
  to: number
};

export type IndexedIssueLocationMessage = {
  flowIndex: number,
  locationIndex: number,
  msg?: string
};

export const issuesByLine = (issues: Array<Issue>) => {
  const index = {};
  issues.forEach(issue => {
    const line = issue.textRange ? issue.textRange.endLine : 0;
    if (!(line in index)) {
      index[line] = [];
    }
    index[line].push(issue);
  });
  return index;
};

export const locationsByLine = (issues: Array<Issue>): { [number]: Array<LinearIssueLocation> } => {
  const index = {};
  issues.forEach(issue => {
    getLinearLocations(issue.textRange).forEach(location => {
      if (!(location.line in index)) {
        index[location.line] = [];
      }
      index[location.line].push(location);
    });
  });
  return index;
};

export const duplicationsByLine = (duplications: Array<*> | null) => {
  if (duplications == null) {
    return {};
  }

  const duplicationsByLine = {};

  duplications.forEach(({ blocks }, duplicationIndex) => {
    blocks.forEach(block => {
      if (block._ref === '1') {
        for (let line = block.from; line < block.from + block.size; line++) {
          if (!(line in duplicationsByLine)) {
            duplicationsByLine[line] = [];
          }
          duplicationsByLine[line].push(duplicationIndex);
        }
      }
    });
  });

  return duplicationsByLine;
};

export const symbolsByLine = (sources: Array<SourceLine>) => {
  const index = {};
  sources.forEach(line => {
    const tokens = splitByTokens(line.code);
    const symbols = flatten(
      tokens.map(token => {
        const keys = token.className.match(/sym-\d+/g);
        return keys != null ? keys : [];
      })
    );
    index[line.line] = symbols.filter(key => key);
  });
  return index;
};
