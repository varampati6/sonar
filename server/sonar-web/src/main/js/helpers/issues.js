/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import { flatten, sortBy } from 'lodash';
import { SEVERITIES } from './constants';
import type { Issue, FlowLocation } from '../components/issue/types';

type TextRange = {
  startLine: number,
  endLine: number,
  startOffset: number,
  endOffset: number
};

type Comment = {
  login: string
};

type User = {
  login: string
};

type RawIssue = {
  assignee?: string,
  author: string,
  comments?: Array<Comment>,
  component: string,
  flows?: Array<{
    locations?: Array<{ msg: string, textRange?: TextRange }>
  }>,
  key: string,
  line?: number,
  project: string,
  rule: string,
  status: string,
  subProject?: string,
  textRange?: TextRange
};

export const sortBySeverity = (issues: Array<*>) =>
  sortBy(issues, issue => SEVERITIES.indexOf(issue.severity));

const injectRelational = (
  issue: RawIssue | Comment,
  source?: Array<*>,
  baseField: string,
  lookupField: string
) => {
  const newFields = {};
  const baseValue = issue[baseField];
  if (baseValue != null && source != null) {
    const lookupValue = source.find(candidate => candidate[lookupField] === baseValue);
    if (lookupValue != null) {
      Object.keys(lookupValue).forEach(key => {
        const newKey = baseField + key.charAt(0).toUpperCase() + key.slice(1);
        newFields[newKey] = lookupValue[key];
      });
    }
  }
  return newFields;
};

const injectCommentsRelational = (issue: RawIssue, users?: Array<User>) => {
  if (!issue.comments) {
    return {};
  }
  const comments = issue.comments.map(comment => {
    const commentWithAuthor = { ...comment, author: comment.login, login: undefined };
    return {
      ...commentWithAuthor,
      ...injectRelational(commentWithAuthor, users, 'author', 'login')
    };
  });
  return { comments };
};

const prepareClosed = (issue: RawIssue) => {
  return issue.status === 'CLOSED' ? { flows: undefined } : {};
};

const ensureTextRange = (issue: RawIssue) => {
  return issue.line && !issue.textRange
    ? {
        textRange: {
          startLine: issue.line,
          endLine: issue.line,
          startOffset: 0,
          endOffset: 999999
        }
      }
    : {};
};

const reverseLocations = (locations: Array<*>) => {
  const x = [...locations];
  x.reverse();
  return x;
};

const splitFlows = (
  issue: RawIssue
  // $FlowFixMe textRange is not null
): { secondaryLocations: Array<FlowLocation>, flows: Array<Array<FlowLocation>> } => {
  const parsedFlows = (issue.flows || [])
    .filter(flow => flow.locations != null)
    // $FlowFixMe flow.locations is not null
    .map(flow => flow.locations.filter(location => location.textRange != null));

  const onlySecondaryLocations = parsedFlows.every(flow => flow.length === 1);

  return onlySecondaryLocations
    ? { secondaryLocations: flatten(parsedFlows), flows: [] }
    : { secondaryLocations: [], flows: parsedFlows.map(reverseLocations) };
};

export const parseIssueFromResponse = (
  issue: Object,
  components?: Array<*>,
  users?: Array<*>,
  rules?: Array<*>
): Issue => {
  const { secondaryLocations, flows } = splitFlows(issue);
  return {
    ...issue,
    ...injectRelational(issue, components, 'component', 'key'),
    ...injectRelational(issue, components, 'project', 'key'),
    ...injectRelational(issue, components, 'subProject', 'key'),
    ...injectRelational(issue, rules, 'rule', 'key'),
    ...injectRelational(issue, users, 'assignee', 'login'),
    ...injectCommentsRelational(issue, users),
    ...prepareClosed(issue),
    ...ensureTextRange(issue),
    secondaryLocations,
    flows
  };
};
