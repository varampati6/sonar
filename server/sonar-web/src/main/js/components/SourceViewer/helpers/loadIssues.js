/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import { searchIssues } from '../../../api/issues';
import { parseIssueFromResponse } from '../../../helpers/issues';

export type Query = { [string]: string };

export type Issues = Array<*>;

// maximum possible value
const PAGE_SIZE = 500;

const buildQuery = (component: string): Query => ({
  additionalFields: '_all',
  resolved: 'false',
  componentKeys: component,
  s: 'FILE_LINE'
});

export const loadPage = (
  query: Query,
  page: number,
  pageSize: number = PAGE_SIZE
): Promise<Issues> => {
  return searchIssues({
    ...query,
    p: page,
    ps: pageSize
  }).then(r =>
    r.issues.map(issue => parseIssueFromResponse(issue, r.components, r.users, r.rules))
  );
};

export const loadPageAndNext = (
  query: Query,
  toLine: number,
  page: number,
  pageSize: number = PAGE_SIZE
): Promise<Issues> => {
  return loadPage(query, page).then(issues => {
    if (issues.length === 0) {
      return [];
    }

    const lastIssue = issues[issues.length - 1];

    if (
      (lastIssue.textRange != null && lastIssue.textRange.endLine > toLine) ||
      issues.length < pageSize
    ) {
      return issues;
    }

    return loadPageAndNext(query, toLine, page + 1, pageSize).then(nextIssues => {
      return [...issues, ...nextIssues];
    });
  });
};

const loadIssues = (component: string, fromLine: number, toLine: number): Promise<Issues> => {
  const query = buildQuery(component);
  return new Promise(resolve => {
    loadPageAndNext(query, toLine, 1).then(issues => {
      resolve(issues);
    });
  });
};

export default loadIssues;
