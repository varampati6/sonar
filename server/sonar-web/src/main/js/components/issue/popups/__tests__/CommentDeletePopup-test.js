/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import React from 'react';
import CommentDeletePopup from '../CommentDeletePopup';
import { click } from '../../../../helpers/testUtils';

it('should render the comment delete popup correctly', () => {
  const onDelete = jest.fn();
  const element = shallow(<CommentDeletePopup onDelete={onDelete} />);
  expect(element).toMatchSnapshot();
  click(element.find('button'));
  expect(onDelete.mock.calls.length).toBe(1);
});
