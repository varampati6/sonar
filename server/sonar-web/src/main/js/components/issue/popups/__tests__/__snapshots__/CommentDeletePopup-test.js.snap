// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the comment delete popup correctly 1`] = `
<BubblePopup
  customClass="bubble-popup-bottom-right"
>
  <div
    className="text-right"
  >
    <div
      className="spacer-bottom"
    >
      issue.comment.delete_confirm_message
    </div>
    <button
      className="button-red"
      onClick={[Function]}
    >
      delete
    </button>
  </div>
</BubblePopup>
`;
